### PR TITLE
Feat: MerkleTree basic functions

### PR DESCRIPTION
**Pull request by LeanSerra**
**This PR**
- Add MerkleTree struct with functions to create the first layer of the tree and to print the tree
- Add type Hash, an array of 32 bytes
- Add `from_leaves` to create a tree from an array of elements that can be transforemed into bytes
- Add `add_element` to add an element to the tree
- Add Display trait to print the tree layers
- Add tests 
  - creating trees with power of two amount of elements
  - creating trees with odd amount of elements
  - creating empty tree
  - creating empty tree and adding elements

### CODE REVIEW

**Review by damiramirez — APPROVED** [review]
LGTM

File: src/merkle_tree.rs [hunk 1]
```diff
@@ -7,34 +9,202 @@ pub struct MerkleTree {
...
+                *last_element = hasher.finalize_reset().into();
             }
-            println!("}}");
+
+            if curr_idx > 0 {
+                curr_idx -= 1;
+            } else {
+                break;
+            }
+            prev_idx -= 1;
```
Comment (dsocolobsky) [inline src/merkle_tree.rs:113]: I would add more comments here because it's hard to follow. Also I think it's clearer if you check for `previous_layer.len() %2==1` instead of substracting 1 and checking for even.

File: src/merkle_tree.rs [hunk 2]
```diff
@@ -7,34 +9,202 @@ pub struct MerkleTree {
...
+
 impl MerkleTree {
     pub fn from_leaves<T: std::convert::AsRef<[u8]>>(leaves: &[T]) -> Self {
         let mut tree = Self::default();
-        tree.build_first_layer(leaves);
+        tree.layers.push(Self::build_first_layer(leaves));
+        while let Some(previous_layer) = tree.layers.first() {
+            if previous_layer.len() <= 1 {
+                break;
+            }
```
Comment (dsocolobsky) [inline src/merkle_tree.rs:32]: Can we do `while let Some(previous_layer) = tree.layers.first() && previous_layer.len() <= 1` here? Not sure if it works

**Review by dsocolobsky — APPROVED** [review]
Approved but I left some comments

File: src/merkle_tree.rs (hunk 2 above)
Comment (LeanSerra) [reply src/merkle_tree.rs:32]: I got ```let` expressions in this position are unstable
see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information``` out of the compiler so I guess not

File: src/merkle_tree.rs (hunk 1 above)
Comment (LeanSerra) [reply src/merkle_tree.rs:113]: Will do on the next PR

File: src/merkle_tree.rs (hunk 2 above)
Comment (dsocolobsky) [reply src/merkle_tree.rs:32]: ok then
